### PR TITLE
osd: advance the pg log pointer during recovery even the object is no…

### PR DESCRIPTION
**Pull request by guangyy**
…t in missing list

If we mark the unfound objects as lost, there is a chance that the missing list in PG log
is empty but the log pointer does not reflect that properly, which result in crash. This
patch move the log pointer manipulate part non-conditional.

Fixes: 13468
Signed-off-by: Guang Yang <yguang@yahoo-inc.com>

### CODE REVIEW

**Comment by guangyy** [comment]
Hi @athanatos , does this look correct?

File: src/osd/PGLog.h [hunk 1]
```diff
@@ -512,11 +512,6 @@ struct PGLog {
...
       missing.got(oid, v);
       
-      // raise last_complete?
-      if (missing.missing.empty()) {
-	log.complete_to = log.log.end();
-	info.last_complete = info.last_update;
-      }
       while (log.complete_to != log.log.end()) {
 	if (missing.missing[missing.rmissing.begin()->second].need <=
 	    log.complete_to->version)
```
Comment (liewegas) [inline src/osd/PGLog.h]: With this change we run teh risk of (less efficiently) running through this loop to advance to the end of the log on the last recovered object.  Perhaps we should leave this hunk behind and just add the second (nearly duplicate) case

Comment (guangyy) [inline src/osd/PGLog.h]: Good catch! Yeah.. I will update the commit.

Other than other, does the analysis make sense for this problem?

**Comment by guangyy** [comment]
Hi @liewegas , updated according to the review comments, please help to review. Thanks!

**Comment by liewegas** [comment]
I think it'll cover the case.  What I'm not sure is whether it is better to patch things up here or deal with it explicitly in the mark unfound code path.. it seems like it might better there.  Is this the recover_got() call in recover_primary() in the case pg_log_entry_t::LOST_REVERT?  Why not put it there and leave recover_got alone?

**Comment by guangyy** [comment]
the recover_got was called with the following stack trace:

2015-10-10 08:13:34.208791 7f20c2c5f700 -1 osd/ReplicatedPG.cc: In function 'void ReplicatedPG::recover_got(hobject_t, eversion_t)' thread 7f20c2c5f700 time 2015-10-10 08:13:34.184066
osd/ReplicatedPG.cc: 9128: FAILED assert(info.last_complete == info.last_update)
ceph version 0.87 (c51c8f9d80fa4e0168aa52685b8de40e42758578)
1: (ReplicatedPG::recover_got(hobject_t, eversion_t)+0x539) [0x873f39]
2: (ReplicatedPG::on_local_recover(hobject_t const&, object_stat_sum_t const&, ObjectRecoveryInfo const&, std::tr1::shared_ptr<ObjectContext>, ObjectStore::Transaction*)+0x2bc) [0x8a8bbc]
3: (ECBackend::handle_recovery_push(PushOp&, RecoveryMessages*)+0xf08) [0x9aa3f8]
4: (ECBackend::handle_message(std::tr1::shared_ptr<OpRequest>)+0x1ae) [0x9b698e]
5: (ReplicatedPG::do_request(std::tr1::shared_ptr<OpRequest>&, ThreadPool::TPHandle&)+0x181) [0x874931]
6: (OSD::dequeue_op(boost::intrusive_ptr<PG>, std::tr1::shared_ptr<OpRequest>, ThreadPool::TPHandle&)+0x178) [0x6505d8]
7: (OSD::ShardedOpWQ::_process(unsigned int, ceph::heartbeat_handle_d*)+0x567) [0x650f37]
8: (ShardedThreadPool::shardedthreadpool_worker(unsigned int)+0x742) [0x9e8a32]
9: (ShardedThreadPool::WorkThreadSharded::entry()+0x10) [0x9ec310]
10: (()+0x7a51) [0x7f20efc9da51]
11: (clone()+0x6d) [0x7f20eec2d9ad]
NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.

**Comment by guangyy** [comment]
That object is removed (coming from the mark found as lost commands):
ReplicatedPG::mark_all_unfound_lost
  ->  PGLog::missing_add_event
        -> pg_missing_t::add_next_event
             -> pg_missing_t::rm

And here we don't check the last_complete pointer.

The other way to fix looks like to update the pointer at pg_missing_t::add_next_event after calling rm?

Thanks @liewegas  for the review.

**Comment by guangyy** [comment]
Hi @liewegas , with the above trace, it looks like doing it from within _recover_got_ is more appropriate for this case? Thanks.

**Comment by liewegas** [comment]
Yeah, I think you're right!

**Comment by liewegas** [comment]
@athanatos can you take a look?

**Comment by liewegas** [comment]
passed testing.

**Comment by athanatos** [comment]
I would like to merge this with a ceph-qa-suite test.

**Comment by liewegas** [comment]
@guangyy still need a teuthology test for this before we can merge

**Comment by athanatos** [comment]
@guangyy Please reach out in #sepia and #ceph-devel if you need help developing a test.

**Comment by guangyy** [comment]
Hi @athanatos @liewegas , sorry for the delayed response, I worked on something else over the last month so a bit delay for this one.

I already have #sepia access, let me add a test case this week.

**Comment by athanatos** [comment]
Sounds good, let us know when the test is ready (link to the ceph-qa-suite PR).

File: src/osd/PGLog.h [hunk 2]
```diff
@@ -511,11 +511,11 @@ struct PGLog {
   void recover_got(hobject_t oid, eversion_t v, pg_info_t &info) {
     if (missing.is_missing(oid, v)) {
       missing.got(oid, v);
-      
+
       // raise last_complete?
       if (missing.missing.empty()) {
-	log.complete_to = log.log.end();
-	info.last_complete = info.last_update;
```
Comment (athanatos) [inline src/osd/PGLog.h]: This looks like an unnecessary whitespace change?

File: src/osd/PGLog.h [hunk 3]
```diff
@@ -511,11 +511,11 @@ struct PGLog {
   void recover_got(hobject_t oid, eversion_t v, pg_info_t &info) {
     if (missing.is_missing(oid, v)) {
       missing.got(oid, v);
-
```
Comment (athanatos) [inline src/osd/PGLog.h]: Unnecessary whitespace change?

**Comment by guangyy** [comment]
Looks like https://github.com/ceph/ceph/pull/6841 is a better fix on root cause, close this one and use that for further tracking.
